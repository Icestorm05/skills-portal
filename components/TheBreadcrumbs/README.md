# The Breadcrumbs

## Functionality

The breadcrumbs component displays a list of profiles divided by a `/`. It is calculated using a profile object passed in, and the Employee ID to find the route to. This component should only be used once within the lifecycle of this application.

## Usage

### Props

`<the-breadcrumbs>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `employee-id` | Number | The Employee ID to find a route to within the Profile object. This will typically be the employee you are currently viewing.
| `profile` | Object | The profile object used to create a route from.
| `route-name` | String | The name of the route to navigate all breadcrumbs to. Expects the route to take an `id` parameter.

### Schemas

The `profile` prop requires the following to exist within the object:

| attribute | type | description
| --- | --- | ---
| `EmployeeId` | Number | The employee ID of the profile.
| `FullName` | String | The full name of the employee. This name will be displayed as the text for the breadcrumb link.
| `children` | Array | An array of further profile objects.

### Example

```html
<the-breadcrumbs :employee-id="3"
                 :profile="profile"
                 route-name="id-profile"/>
```

```ts
import Component from 'nuxt-class-component';
import Vue from 'vue';

import TheBreadcrumbs from '~/components/TheBreadcrumbs/TheBreadcrumbs.vue';

@Component({
    components: {
        TheBreadcrumbs,
    },
})
export default class GenericComponent extends Vue {
    private profile = {
        EmployeeId: 1,
        FullName: 'John Doe',
        children: [{
            EmployeeId: 2,
            FullName: 'Jane Doe',
        }, {
            EmployeeId: 3,
            FullName: 'Jake Doe',
        }]
    }
}
```