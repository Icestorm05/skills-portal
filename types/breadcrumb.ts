import { Location } from 'vue-router';

export interface IBreadcrumb {
    link: Location;
    text: string;
}