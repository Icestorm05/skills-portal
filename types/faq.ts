export interface IFaq {
    question: string;
    answers: IFaqAnswer[];
}

export interface IFaqAnswer {
    type: 'p' | 'b' | 'ul' | 'table';
    text: string | string[] | string[][];
}