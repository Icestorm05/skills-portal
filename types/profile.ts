export interface IProfile {
    Biography: string;
    Company: string;
    Director: string;
    EmailAddress: string;
    EmployeeId: number;
    FirstLineReporting: string;
    FullName: string;
    JobTitle: string;
    KnownAs: string;
    MainStaffNumber: string;
    SecondLineReporting: string;
    Surname: string;
    children: IProfile[];
}