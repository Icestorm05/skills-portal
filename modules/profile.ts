import { IProfile } from "types/profile";

/**
 * Calculates how deep a profile is in a hierarchy of profiles.
 * @param profiles {IProfile[]} The hierarchy of profiles.
 * @param profileToFind {IProfile} The profile to find.
 * @param level {number?} The depth level when no profile is found. Defaults to 0.
 * @return {number} The depth level.
 */
export const deepLevel = (
  profiles: IProfile[],
  profileToFind: IProfile,
  level: number = 0
): number => {
  return profiles.reduce((depth, profile: IProfile) => {
    if (profile.EmployeeId === profileToFind.EmployeeId) {
      return level + 1;
    } else if (profile.children) {
      const deepDepth = deepLevel(profile.children, profileToFind, level + 1);
      if (deepDepth) {
        return deepDepth;
      }
    }
    return depth;
  }, 0);
};

/**
 * Searches a hierarchy of profiles for one that matches a specified employee ID.
 * @param profiles {IProfile[]} The profile hierarchy.
 * @param id {number} The employee ID to find.
 * @return {IProfile | null} The matched profile, or null if not found.
 */
export const findProfile = (
  profiles: IProfile[],
  id: number
): IProfile | null => {
  return profiles.reduce((matchedProfile: IProfile | null, currProfile) => {
    if (!matchedProfile) {
      if (currProfile.EmployeeId === id) {
        return currProfile;
      } else if (currProfile.children) {
        return findProfile(currProfile.children, id);
      } else {
        return null;
      }
    }
    return matchedProfile;
  }, null);
};
