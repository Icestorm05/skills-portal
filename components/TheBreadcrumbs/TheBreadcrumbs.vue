<template lang="pug">
    v-breadcrumbs(divider="/")#TheBreadcrumbs
        v-breadcrumbs-item(nuxt
                           :to="{ name: routeName }").BreadcrumbItem {{ profile.FullName }}
        v-breadcrumbs-item(v-for="route in path(profile.children)"
                           :key="route.text"
                           nuxt
                           :to="route.link").BreadcrumbItem {{ route.text }}
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import Vue from "vue";
import { Prop } from "vue-property-decorator";

import { IBreadcrumb } from "types/breadcrumb";
import { IProfile } from "types/profile";

import VBreadcrumbs, {
  VBreadcrumbsItem
} from "vuetify/es5/components/VBreadcrumbs";

@Component({
  components: {
    VBreadcrumbs,
    VBreadcrumbsItem
  }
})
export default class TheBreadcrumbs extends Vue {
  @Prop({ type: Number, required: true })
  private employeeId: number;
  @Prop({ type: Object, required: true })
  private profile: IProfile;
  @Prop({ type: String, required: true })
  private routeName: string;

  private path(childProfiles: IProfile[]): IBreadcrumb[] {
    return childProfiles.reduce((path: IBreadcrumb[], childProfile) => {
      if (this.employeeId === childProfile.EmployeeId) {
        path.push({
          link: {
            name: this.routeName,
            params: { id: childProfile.EmployeeId.toString() }
          },
          text: childProfile.FullName
        });
      } else {
        const deepProfiles = this.path(childProfile.children);
        if (deepProfiles.length) {
          path.push({
            link: {
              name: this.routeName,
              params: { id: childProfile.EmployeeId.toString() }
            },
            text: childProfile.FullName
          });
        }
        return path.concat(deepProfiles);
      }
      return path;
    }, []);
  }
}
</script>
