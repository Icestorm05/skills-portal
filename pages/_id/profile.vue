<template lang="pug">
    #ProfilePage
        user-profile(:profile="profile"
                     :disabled="inProgress"
                     :editable="currentDepth === 1"
                     @updated-biography="updateBiography")
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import Vue from "vue";
import { Getter, State } from "vuex-class";

import { IProfile } from "types/profile";

import UserProfile from "~/components/UserProfile/UserProfile.vue";

@Component({
  components: {
    UserProfile
  }
})
export default class ProfilePage extends Vue {
  @Getter("profile/current") private profile: IProfile;
  @Getter("profile/currentDepth") private currentDepth: number;

  @Getter("progress/isVisible") private inProgress: boolean;

  private async updateBiography(Biography: string) {
    try {
      this.$store.commit("progress/show");
      await this.$store.dispatch("profile/edit", {
        ...this.profile,
        Biography
      });
      this.$store.commit("snackbar/show", {
        message: "Successfully updated bio.",
        color: "success"
      });
    } catch (err) {
      this.$store.commit("snackbar/show", {
        message: "Failed to update bio.",
        color: "error"
      });
    } finally {
      this.$store.commit("progress/hide");
    }
  }
}
</script>
