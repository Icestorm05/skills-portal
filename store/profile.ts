import { IProfile } from "types/profile";
import { deepLevel, findProfile } from "~/modules/profile";

const defaultProfile = {
  Biography: "",
  Director: "",
  EmailAddress: "",
  EmployeeId: 0,
  FirstLineReporting: "",
  FullName: "",
  JobTitle: "",
  KnownAs: "",
  SecondLineReporting: "",
  Surname: "",
  children: []
};

export const state = () => ({ ...defaultProfile });

export const mutations = {
  set(currentProfile: IProfile, profile: IProfile) {
    Object.keys(currentProfile).forEach(
      key => (currentProfile[key] = profile[key])
    );
  },
  updateBiography(currentProfile: IProfile, biography: string) {
    currentProfile.Biography = biography;
  },
  reset(currentProfile: IProfile) {
    Object.keys(currentProfile).forEach(
      key => (currentProfile[key] = defaultProfile[key])
    );
  }
};

export const actions = {
  async get({ commit }) {
    const data = await this.$axios.$get("/profile");
    commit("set", data.profile);
  },
  async edit({ commit }, profile: IProfile) {
    const url = `/employees/${profile.EmployeeId}`;
    await this.$axios.$put(url, { employee: profile });
    commit("updateBiography", profile.Biography);
  }
};

export const getters = {
  current(currentProfile: IProfile, localGetters, globalState) {
    const id = globalState.route
      ? parseInt(globalState.route.params.id, 10)
      : 0;
    const profile = findProfile(
      [currentProfile],
      id || currentProfile.EmployeeId
    );
    return profile ? profile : currentProfile;
  },
  currentDepth(currentProfile: IProfile, localGetters) {
    return localGetters.current.EmployeeId
      ? deepLevel([currentProfile], localGetters.current)
      : 0;
  },
  hasTeam(currentProfile: IProfile, localGetters) {
    return localGetters.current.children.length ? true : false;
  }
};
