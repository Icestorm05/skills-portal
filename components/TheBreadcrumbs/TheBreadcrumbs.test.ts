import TheBreadcrumbs from "./TheBreadcrumbs.vue";
import { mount } from "@vue/test-utils";

import { IProfile } from "types/profile";

declare const expect: jest.Expect;

const profile: IProfile = {
  Biography: "",
  Company: "",
  Director: "",
  EmailAddress: "",
  EmployeeId: 1,
  FirstLineReporting: "",
  FullName: "John Doe",
  JobTitle: "",
  KnownAs: "",
  MainStaffNumber: "",
  SecondLineReporting: "",
  Surname: "",
  children: []
};

const profile2: IProfile = {
  ...profile,
  EmployeeId: 2,
  FullName: "Jane Doe"
};

const profile3: IProfile = {
  ...profile,
  EmployeeId: 3,
  FullName: "Jake Doe"
};

const profileWithTwoLevels: IProfile = {
  ...profile,
  children: [profile2]
};

const profileWithThreeLevels: IProfile = {
  ...profile,
  children: [
    {
      ...profile2,
      children: [profile3]
    }
  ]
};

const profileWithManyChildren: IProfile = {
  ...profile,
  children: [profile2, profile3]
};

const stubs = {
  "nuxt-link": "<a></a>"
};

describe("The Breadcrumbs", () => {
  test("is a Vue instance", () => {
    const wrapper = mount(TheBreadcrumbs, {
      stubs,
      propsData: { employeeId: 0, profile, routeName: "id-team" }
    });
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("snapshot has not changed", () => {
    const wrapper = mount(TheBreadcrumbs, {
      stubs,
      propsData: {
        employeeId: 3,
        profile: profileWithThreeLevels,
        routeName: "id-team"
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("displays profile full name as first breadcrumb", () => {
    const wrapper = mount(TheBreadcrumbs, {
      stubs,
      propsData: { employeeId: 0, profile, routeName: "id-team" }
    });
    expect(
      wrapper.find("#TheBreadcrumbs .BreadcrumbItem .breadcrumbs__item").text()
    ).toEqual(profile.FullName);
  });

  test("does not display second breadcrumb with invalid employee id to find in hierarchy", () => {
    const wrapper = mount(TheBreadcrumbs, {
      stubs,
      propsData: {
        employeeId: 0,
        profile: profileWithTwoLevels,
        routeName: "id-team"
      }
    });
    const breadcrumbs = wrapper.findAll(
      "#TheBreadcrumbs .BreadcrumbItem .breadcrumbs__item"
    );
    expect(breadcrumbs.length).toEqual(1);
    expect(breadcrumbs.at(0).text()).toEqual(profile.FullName);
  });

  test("displays second breadcrumb with valid employee id to find in hierarchy", () => {
    const wrapper = mount(TheBreadcrumbs, {
      stubs,
      propsData: {
        employeeId: 2,
        profile: profileWithTwoLevels,
        routeName: "id-team"
      }
    });
    const breadcrumbs = wrapper.findAll(
      "#TheBreadcrumbs .BreadcrumbItem .breadcrumbs__item"
    );
    expect(breadcrumbs.length).toEqual(2);
    expect(breadcrumbs.at(0).text()).toEqual(profile.FullName);
    expect(breadcrumbs.at(1).text()).toEqual(profile2.FullName);
  });

  test("displays second breadcrumb with valid employee id to find in many child hierarchy", () => {
    const wrapper = mount(TheBreadcrumbs, {
      stubs,
      propsData: {
        employeeId: 3,
        profile: profileWithManyChildren,
        routeName: "id-team"
      }
    });
    const breadcrumbs = wrapper.findAll(
      "#TheBreadcrumbs .BreadcrumbItem .breadcrumbs__item"
    );
    expect(breadcrumbs.length).toEqual(2);
    expect(breadcrumbs.at(0).text()).toEqual(profile.FullName);
    expect(breadcrumbs.at(1).text()).toEqual(profile3.FullName);
  });

  test("displays third breadcrumb with valid employee id to find in hierarchy", () => {
    const wrapper = mount(TheBreadcrumbs, {
      stubs,
      propsData: {
        employeeId: 3,
        profile: profileWithThreeLevels,
        routeName: "id-team"
      }
    });
    const breadcrumbs = wrapper.findAll(
      "#TheBreadcrumbs .BreadcrumbItem .breadcrumbs__item"
    );
    expect(breadcrumbs.length).toEqual(3);
    expect(breadcrumbs.at(0).text()).toEqual(profile.FullName);
    expect(breadcrumbs.at(1).text()).toEqual(profile2.FullName);
    expect(breadcrumbs.at(2).text()).toEqual(profile3.FullName);
  });
});
