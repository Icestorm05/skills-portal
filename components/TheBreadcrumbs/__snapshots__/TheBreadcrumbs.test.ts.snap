// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Breadcrumbs snapshot has not changed 1`] = `
<ul class="breadcrumbs" id="TheBreadcrumbs">
    <li class="BreadcrumbItem">
        <nuxt-link class="breadcrumbs__item">John Doe</nuxt-link>
    </li>
    <li class="breadcrumbs__divider">/</li>
    <li class="BreadcrumbItem">
        <nuxt-link class="breadcrumbs__item">Jane Doe</nuxt-link>
    </li>
    <li class="breadcrumbs__divider">/</li>
    <li class="BreadcrumbItem">
        <nuxt-link class="breadcrumbs__item">Jake Doe</nuxt-link>
    </li>
</ul>
`;
